RFC 008 - Release Versioning (#17)
thegooddocsproject/request-for-comment: RFC-008-release-versioning.md

# RFC008 - Release Versioning Strategy

## Proposed by

Aaron Peters

## Current status

See the status indicated on the pull request description.

## Proposal overview

As The Good Docs Project moves towards a formal release structure, this RFC proposes a standard versioning scheme to apply at least to the project's Product Outputs, as defined by outputs directly employed by users in their own documentation efforts. For example, this includes the collection of templates (which users copy as a starting point for their own documents), but not whitepapers (which serve to inform their audience, but are not used verbatim).

> TGDP will use the Semantic Versioning convention for the versioning of individual Product Outputs (at the time of writing, this includes the template collection and the DocTools website package).

Reviewer: Nitpick suggestion: Apply [timeless documentation](https://developers.google.com/style/timeless-documentation) principles. Someone reading this in a year's time won't know when this doc was written.
```suggestion
TGDP will use the Semantic Versioning convention for the versioning of individual Product Outputs (as at November 2021, this includes the template collection and the DocTools website package).
```

> :triangular_flag_on_post: For the purposes of this RFC, "features" and "functionality" can refer to new **content**, such as additional sections or boilerplate text, added to doc templates.

## Motivation

> Using the well-established [https://semver.org/](Semantic Versioning) framework will provide the following benefits:

Reviewer: Markdown bug:
```suggestion
Using the well-established [Semantic Versioning](https://semver.org/) framework will provide the following benefits:
```

- It will demonstrate the project's progress to users and prospective users through publicly visible notifications.
- It will set expectations to current users regarding the extent of change(s) contained in a newly-release version of a Product Output, as compared to the version the user currently has, including potentially breaking backward-compatibility.
- It provides contributors to the project a common point of reference when collaborating, making the team more efficient.
- It provides downstream integrators of TGDP's Product Outputs an easy way to identify if their own work products will need to be updated to support new features, or if there is a risk of compatibility issues.

> ## Proposal

Reviewer: For the content that is in this proposal section, I think it is a good foundation, and is pretty close to right. I also think we should extend it to cover:
* From a general documentation point of view (say for docs which describe a product), what constitutes a sematic update based on major.minor.patch. What is:
  * A 1.0 release
  * A major update
  * A minor update
  * A patch update
* Same questions, but this time for our template.
* I think we should also discuss the difference between a document update, and a documentation set update. (Analogy is a package version vs a linux distribution version.)

Reviewer: Okay, I've [drafted a proposal for how semantic versioning should apply to docs](https://docs.google.com/document/d/1gPr_OHoUkg5Wjj3UWYX78ORtTBArSJck4mWrKhWpfPA/edit#).

Author: I made a major mistake and merged this RFC in, but I've noted your comments in the "Feedback" section of the new RFC that I'm about to open. I'll let @acpkendo  decide if he wants to move it up into the body of the RFC.

## Dependencies

This RFC should be considered together with RFC-006, which describes the high-level planning process to be used for the project's Releases.

## Links and prior art

- [Meeting notes from discussion on versioning](https://docs.google.com/document/d/1Oc3hNhJRrXsr-abEmUKAdamC5aKJydtldvpcSK_Hv_E/edit)

## Open questions

- Should we limit Product Outputs to one per Working Group? For example, the Community Docs Working Group would produce the "Template Collection," as opposed to considering each template as a separate Product Output?
  - In the context of this RFC, this means the collection as a whole would be versioned together.
  - A result of this would be that, as defined by the SemVer standard, **any** update to a template that adds materially to it (i.e. not fixing typos or other "bugs") would mean an increase in the minor version of the collection as a whole.
- How long will the project aim to support legacy major versions? Minor versions?

> 

Author: Question (non-blocking): Do we need to worry about what to call releases that are in progress? Like will we use a code name to refer to the release in the community? If so, I propose we use an alphabetic naming scheme (first code name release is A, second code name release is B) and then name each code release after an author starting with that name, such as Austen, Baldwin, Camus, etc. We can pick the next 5 release names at a time through community vote possibly.

I don't want this to be a contentious suggestion so I will gladly withdraw it if it stirs up too much controversy. I don't feel terribly strongly about it. I just feel that code names for releases are easier to work with than numbers.

Reviewer 2: 💯  I've done inventor(s) and my-little ponies 😁  as an alternative to authors and agreed when planning and referring to a release milestone names work much better for a memory-retention perspective.

Reviewer: As per poll set up by @barbaricyawps , naming convention was to use bridge names.

Author: Yes, agreed. Let's make sure we're more specific here in the RFC to reduce confusion. My suggested wording:

As discussed in the January 26, 2022 general US meeting where a quorum of project steering committee members were present, we will adopt internal codenames to refer to our template and tool releases. We ran an internal community poll to determine what the naming convention should be and the decision was to name after notable bridges throughout the world. Our codenames are alphabetical, which means that the first release will feature a bridge name that starts with the letter A. For example: Accademia, Ada, Akashi, Akrobaten, Alamillio, Anzac. The following release will have a bridge name that starts with the letter B and so forth. Bridge names for each code release will be determined by the PSC or through informal community polls.

Author: I made a major mistake and accidentally merged this RFC in prematurely. I'm opening up a new one. I've added this suggested paragraph to a new section of the RFC.

# RFC008 - Release Versioning Strategy

## Proposed by

Aaron Peters

## Current status

See the status indicated on the pull request description.

## Proposal overview

> As The Good Docs Project moves towards a formal release structure, this RFC proposes a standard versioning scheme to apply at least to the project's Product Outputs, as defined by outputs directly employed by users in their own documentation efforts. For example, this includes the collection of templates (which users copy as a starting point for their own documents), but not whitepapers (which serve to inform their audience, but are not used verbatim).

Reviewer: While there is less of a need to use versioning for whitepapers (and other documents), I can think of a few use cases where I might use semantic versioning for white papers. Especially if there are multiple editors/reviewers of the white paper. So maybe we should drop examples of where semantic versioning is not used, and leave it to personal choice. (If you really want a non-versioning article, I suggest referring to a news article.)
```suggestion
As The Good Docs Project moves towards a formal release structure, this RFC proposes a standard versioning scheme to apply at least to the project's Product Outputs, as defined by outputs directly employed by users in their own documentation efforts. For example, this includes the collection of templates (which users copy as a starting point for their own documents).
```

Author: I made a major mistake and accidentally merged this RFC in prematurely. I added your suggestion to the body of the RFC.

1. The first such tag within a Release cycle should also carry the "-alpha" suffix.
1. Further work during the release will focus on stabilizing the existing functionality and making it ready for general/production use.
1. Once the Product Output is deemed ready, it will be tagged with the same semantic version number and the "-beta" suffix.
1. At this point it should be released to a community of beta testers to actually use the Product Output. Feedback from this use should be actively solicited and captured.
1. Provided there are no "blocker"-level issues issues remaining with the Product Output as the Release's target date approaches, the Release Manager will coordinate the completion of any final updates before signaling for them to be tagged with the same semantic version number, minus any suffix (indicating a gamma/final release).
1. Once tagged, building and packaging the Product Outputs can commence.
2. Following the gamma release of an Output, "patch" releases may be released in parallel to updated versions, as described in the section Semantic versioning addendum for documents below.

## Semantic versioning addendum for documents

> 

Reviewer: Suggestion: This section is based on an external source document which I'm expecting to be getting updated by the documentation and developer community. (It currently has a bunch of review comments which are yet to be actioned).
As such, it may be helpful to reference the point of truth, and help readers understand whether this RFC content is the latest version or not.
```suggestion

Source: [Semantic versioning addendum for documents. Version 0.1-alpha.4. January 2020](https://docs.google.com/document/d/1gPr_OHoUkg5Wjj3UWYX78ORtTBArSJck4mWrKhWpfPA).

```

Author: I made a major mistake and accidentally merged this RFC in prematurely. I've added this suggestion to the body of the RFC.

-   [Semantic Versioning and Structure for IETF Specifications, draft-claise-semver-02](https://datatracker.ietf.org/doc/html/draft-claise-semver-02).
-   [Open Geospatial Consortium's Policy Directive 18](https://portal.ogc.org/public_ogc/directives/directives.php) defines use of Major.Minor.Bug-fix versioning for spatial specifications and related documents.

## Consequences

This decision will dictate how Releases are communicated to both internal and external parties. Using the Semantic Versioning convention will be familiar to many of the project's most important audiences, namely open source projects (many of which use the same or similar numbering schemes) and technical writers of various levels of experience.

## Dependencies

This RFC should be considered together with RFC-006, which describes the high-level planning process to be used for the project's Releases.

## Links and prior art

- [Meeting notes from discussion on versioning](https://docs.google.com/document/d/1Oc3hNhJRrXsr-abEmUKAdamC5aKJydtldvpcSK_Hv_E/edit)

> 

Reviewer: Suggestion: Probably worth referencing the point of truth from here too.

```suggestion
- [Semantic versioning addendum for documents. Version 0.1-alpha.4. January 2020](https://docs.google.com/document/d/1gPr_OHoUkg5Wjj3UWYX78ORtTBArSJck4mWrKhWpfPA).

```

Author: I made a major mistake and accidentally merged this RFC in prematurely. I've added your suggestion to the body of the new RFC I'm opening.
